### ISSUE TITLE
New modules and settings to make newman more flexible for both manual and automated runs

### ISSUE THREAD

**Issue by correajd** [postmanlabs/newman#1065]
Details
-->
 1. Newman Version 3.5.2
 2. OS details Mac OS Sierra and Jenkins running Debian
 3. Newman Library
 3. Not a bug
 4. Desired not expected
 5. N/A
 6. Sample collection, and auxilliary files (minus the sensitive details):
 7. Screenshots (if applicable):

Hello, my team is starting to use newman for automated api testing. We are excited about the opportunity to use this for our developers, manual QA, and automation.

We are particularly excited about the data driven aspect of running newman either through the runner or the CLI. 

The problem is that when we use tv4 validations and the data from the csv/json file, we notice our test module gets very cluttered with spaghetti code. 

Take a look at the following code. 

```javascript
/* The expected code should be a pre condition that i can set, something like a dropdown.
Ideally this would be an option near the pre-request script and test script
If this doesn't pass then the "Tests" module wouldn't need to be executed. 
*/
let responseCodeTest = (responseCode.code === 200);
tests['Status code is 200'] = responseCodeTest;

if(responseCodeTest){
    let body = JSON.parse(responseBody);
    if(typeof body == 'object'){
       /* Again this should be in another TV4 module where I can set the expected schema before i do actual validations on the data itself, i first want to check the presents of the data before I test the quality */
        let schema = getSchema();
        let schemaMatch = tv4.validate([body], schema);
        /*  go down to the validations method */
        (schemaMatch ? validations(data, body) : retry());
        postman.setEnvironmentVariable("customerId", body.personId);
        postman.setEnvironmentVariable("scheduleId", body.id);
    } else {
        retry();
    }
} else {
    stop()
}

function validations(data, body) {
    tests['object matches data schema'] = true;
    /* again this should be in a separate module next to the "Pre-request Script" and "Tests"
    that would only be executed if this were being run in the Runner, CLI, or newman code
    This would help clean code up quite a bit
 Perhaps call it "data driven tests"? */
    if(data.merchantID){
        tests["source stored"] = (data.source.toLowerCase() === body.source.toLowerCase());
        tests["referrer stored"] = (data.referrer.toLowerCase() === body.referrer.toLowerCase());
        tests["campaign code stored"] = (data.campaignCode === body.campaign);
    }
}

function stop() {
    postman.setNextRequest(null);
}

var requestName = this.request.name;
/* sometimes i don't get an immediate schema validation success when validating the response because the state of the object my not be complete yet,
It would be nice to have a built in function for retries with a retry attempt limit and custom wait option */
function retry() {
    console.log("Validation failed: ", tv4.error);
    wait(5000);
    postman.setNextRequest('GET Recurring Payment Definition');
}

function wait(millis) {
    var date = new Date();
    var curDate = null;
    do {
        curDate = new Date();
    } while( curDate-date < millis);
}

function getSchema() {
    return {
        "items": {
            "type": "object",
            "properties": {
                "id": {"type": "string"},
                "customerId": {"type": "string"},
                "startDate": {"type": "string"},
                "paymentType": {"type": "string"},
                "serviceType": {"type": "string"},
                "status": {"type": "string"},
                "price": {"type": "number"},
                "monthlyPrice": {"type": "number"},
                "stopBilling": {"enum": ["Y","N"]},
                "packageId": {"type": "string"},
                "isRecurring": {"enum": ["Y","N"]},
                "priceId": {"type": "string"},
                "campaign": {"type": "string"},
                "isFinanced": {"enum": ["Y","N"]},
                "overridingPaymentMethod": {"type": "string"},
                "installments": {"type": "number"}
            }
        }
    };
}
```
To summarize, things would be more flexible if we had:

Precondition 1: Valid response code option
Precondition 2: Schema validation
Precondition 3: Existence of data which would then execute the "Data Tests" module code if the tests were being run in automation
Option 1: Customize retries instead of using postman.setNextRequest("name" || null);
Option 2: Pull data files into manual runs

**Comment by shamasis** [comment]
This should be resolved by better TV4 assertion APIs we are planning in Postman Sandbox.

**Comment by codenirvana** [comment]
The latest Postman sandbox adds support for [AJV](https://github.com/ajv-validator/ajv) and deprecates support for TV4.
